src/MiddlewareDispatcher.php

```diff
@@ -13,88 +13,89 @@
+    private RequestHandlerInterface $nextHandler;
+    private ContainerInterface $container;
 
     /**
-     * @var ContainerInterface
+     * Contains a chain of middleware wrapped in handlers.
+     * Each handler points to the handler of middleware that will be processed next.
+     * @var RequestHandlerInterface|null stack of middleware
      */
-    private $container;
+    private ?RequestHandlerInterface $stack = null;
 
     public function __construct(
-        array $middlewares,
         ContainerInterface $container,
         RequestHandlerInterface $nextHandler = null
     ) {
-        if ($middlewares === []) {
-            throw new \InvalidArgumentException('Middlewares should be defined.');
-        }
-
         $this->container = $container;
 
-        foreach ($middlewares as $middleware) {
-            $this->add($middleware);
-        }
-
         $responseFactory = $container->get(ResponseFactoryInterface::class);
 
         $this->nextHandler = $nextHandler ?? new NotFoundHandler($responseFactory);
     }
 
-    private function addCallable(callable $callback): void
-    {
-        $this->middlewares[] = new Callback($callback, $this->container);
-    }
-
-    public function add($middleware): void
+    /**
+     * @param callable|MiddlewareInterface $middleware
+     * @return self
+     */
+    public function addMiddleware($middleware): self
     {
         if (is_callable($middleware)) {
-            $this->addCallable($middleware);
-        } elseif ($middleware instanceof MiddlewareInterface) {
-            $this->middlewares[] = $middleware;
-        } else {
+            $middleware = new Callback($middleware, $this->container);
+        }
+
+        if (!$middleware instanceof MiddlewareInterface) {
             throw new \InvalidArgumentException('Middleware should be either callable or MiddlewareInterface instance. ' . get_class($middleware) . ' given.');
         }
+
+        array_unshift($this->middlewares, $middleware);
+
+        return $this;
     }
 
     public function dispatch(ServerRequestInterface $request): ResponseInterface
     {
-        reset($this->middlewares);
-        return $this->handle($request);
+        return $this->process($request, $this->nextHandler);
     }
 
-    /**
-     * @internal Please use {@see dispatch()} or {@see process()} instead
-     * @param ServerRequestInterface $request
-     * @return ResponseInterface
-     */
-    public function handle(ServerRequestInterface $request): ResponseInterface
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
     {
-        $middleware = current($this->middlewares);
-        next($this->middlewares);
-        if ($middleware === false) {
-            if ($this->nextHandler !== null) {
-                return $this->nextHandler->handle($request);
+        if ($this->stack === null) {
+            for ($i = count($this->middlewares) - 1; $i >= 0; $i--) {
```

Comment:
```suggestion
            foreach ($this->middlewares as $middleware) {
```

Reply:
I thought exactly the same but I so lazy...